pdns/recursordist/rec-xfr.cc

```diff
@@ -0,0 +1,495 @@
+          auto groupName = txt->d_text;
+          groupName = unquotify(groupName);
+          auto iter = params.d_defaults.find(groupName);
+          if (iter == params.d_defaults.end()) {
+            logger->info(Logr::Debug, "No match for group in YAML config", "name", Logging::Loggable(name), "groupName", Logging::Loggable(groupName));
+            continue;
+          }
+          logger->info(Logr::Debug, "Match for group in YAML config", "name", Logging::Loggable(name), "groupName", Logging::Loggable(groupName));
+          defsIter = iter;
+          break;
+        }
+      }
+      if (defsIter == params.d_defaults.end()) {
+        logger->info(Logr::Error, "No match for group in YAML config", "name", Logging::Loggable(name));
+        continue;
+      }
+      auto target = ptr->getContent();
+      pdns::rust::settings::rec::ForwardZone forward;
+      forward.zone = target.toString();
+      forward.recurse = defsIter->second.recurse;
+      forward.notify_allowed = defsIter->second.notify_allowed;
+      for (const auto& value : defsIter->second.forwarders) {
+        forward.forwarders.emplace_back(value);
+      }
+      forward.validate("catz");
+      forwards.emplace_back(std::move(forward));
+    }
+  }
+
+  // Simple approach: just replace everything. Keeping track of changes to members is a bit involved as the
+  // members themselves can change, be added ot deleted, but also the group records.
+  pdns::rust::settings::rec::api_delete_zones(zonesFile);
+  pdns::rust::settings::rec::api_add_forward_zones(zonesFile, forwards);
+  reloadZoneConfiguration(true);
+}
+
+bool CatalogZone::versionCheck() const
+{
+  auto records = d_records.equal_range(std::make_pair(cVersion, QType::TXT));
+  bool invalid = false;
+  size_t count = 0;
+  for (auto record = records.first; record != records.second; ++record) {
+    const auto txt = getRR<TXTRecordContent>(record->second);
+    if (txt == nullptr) {
+      invalid = true;
+      continue;
+    }
+    auto str = txt->d_text;
+    if (str != "\"2\"") {
+      invalid = true;
+      continue;
+    }
+    ++count;
+  }
+  return !invalid && count == 1;
+}
+
+bool CatalogZone::dupsCheck() const
+{
+  std::unordered_set<DNSName> values;
+  bool invalid = false;
+  for (const auto& [key, record] : d_records) {
+    if (key.second != QType::PTR) {
+      continue;
+    }
+    const auto ptr = getRR<PTRRecordContent>(record);
+    if (ptr == nullptr) {
+      invalid = true;
+      continue;
+    }
+    if (!values.emplace(ptr->getContent()).second) {
+      invalid = true;
+      break;
+    }
+  }
+  return !invalid;
+}
+
+static shared_ptr<const SOARecordContent> loadZoneFromServer(Logr::log_t plogger, const ComboAddress& primary, const DNSName& zoneName, shared_ptr<CatalogZone>& zone, const TSIGTriplet& tsigTriplet, size_t maxReceivedBytes, const ComboAddress& localAddress, uint16_t axfrTimeout)
```

Comment:
Not an issue but the name and the first log lines feel like this function is intended to load generic zones, not only catalog zones, then by the end we have catalog-specific processing which surprised me (but is fine).

Reply:
The class started with the goal to use a single (base or template) class for both RPZ and catz transfers, but it turned out the goal was a bit ambitious. Some day I might unify the common code. What you are seeing is remains of the original goal. 

The subsystem used with the log messages is "fwcatzixfr", so even though the log message is generic, the subsystem value makes the purpose clear I think.

Reply:
Fine by me, then :)